lib/Dialect/CGGI/Transforms/BooleanLineVectorizer.cpp

```diff
@@ -0,0 +1,224 @@
+      lhs->getResultTypes() != rhs->getResultTypes() ||
+      lhs->getAttrs() != rhs->getAttrs()) {
+    return false;
+  }
+  return OpTrait::hasElementwiseMappableTraits(lhs);
+}
+
+bool tryBoolVectorizeBlock(Block *block, MLIRContext &context) {
+  graph::Graph<Operation *> graph;
+  for (auto &op : block->getOperations()) {
+    if (!op.hasTrait<OpTrait::Elementwise>()) {
+      continue;
+    }
+
+    graph.addVertex(&op);
+    SetVector<Operation *> backwardSlice;
+    BackwardSliceOptions options;
+    options.omitBlockArguments = true;
+
+    getBackwardSlice(&op, &backwardSlice, options);
+    for (auto *upstreamDep : backwardSlice) {
+      // An edge from upstreamDep to `op` means that upstreamDep must be
+      // computed before `op`.
+      graph.addEdge(upstreamDep, &op);
+    }
+  }
+
+  if (graph.empty()) {
+    return false;
+  }
+
+  auto result = graph.sortGraphByLevels();
+  assert(succeeded(result) &&
+         "Only possible failure is a cycle in the SSA graph!");
+  auto levels = result.value();
+
+  LLVM_DEBUG({
+    llvm::dbgs()
+        << "Found operations to vectorize. In topo-sorted level order:\n";
+    int level_num = 0;
+    for (const auto &level : levels) {
+      llvm::dbgs() << "\nLevel " << level_num++ << ":\n";
+      for (auto op : level) {
+        llvm::dbgs() << " - " << *op << "\n";
+      }
+    }
+  });
+
+  bool madeReplacement = false;
+  for (const auto &level : levels) {
+    DenseMap<Operation *, SmallVector<Operation *, 4>> compatibleOps;
+    for (auto *op : level) {
+      bool foundCompatible = false;
+      for (auto &[key, bucket] : compatibleOps) {
+        if (areCompatibleBool(key, op)) {
+          compatibleOps[key].push_back(op);
+          foundCompatible = true;
+        }
+      }
+      if (!foundCompatible) {
+        compatibleOps[op].push_back(op);
+      }
+    }
+    LLVM_DEBUG(llvm::dbgs()
+               << "Partitioned level of size " << level.size() << " into "
+               << compatibleOps.size() << " groups of compatible ops\n");
+
+    // Loop over all the compatibleOp groups
+    // Each loop will have the key and a bucket with all the operations in
+    for (auto &[key, bucket] : compatibleOps) {
+      if (bucket.size() < 2) {
+        continue;
+      }
+
+      LLVM_DEBUG({
+        llvm::dbgs() << "[START] Bucket \t Vectorizing ops:\n";
+        for (auto op : bucket) {
+          llvm::dbgs() << " - " << *op << "\n";
+        }
+      });
+
+      OpBuilder builder(bucket.back());
+      // relies on CGGI ops having a single result type
+      Type elementType = key->getResultTypes()[0];
+      RankedTensorType tensorType = RankedTensorType::get(
+          {static_cast<int64_t>(bucket.size())}, elementType);
+
+      SmallVector<Value, 4> vectorizedOperands;
+      SmallVector<StringAttr, 4> vectorizedGateOperands;
+
+      for (auto op : bucket) {
+        std::string str;
+        if (isa<cggi::AndOp>(op)) {
```

Comment:
nit: better to use `TypeSwitch`

```
  std::string str = llvm::TypeSwitch<Operation &, std::string>(op)
      .Case<cggi::AndOp>([](auto op) { return std::string("and"); })
      .Case<cggi::NandOp>([](auto op) { return std::string("nand"); })
      ...
      .Default([&](Operation &) { return "unknown"; });
```

Alternatively, could probably do string munging on the operation name.